OneSignalSDK/onesignal/src/main/java/com/onesignal/NotificationLimitManager.java

```diff
@@ -0,0 +1,126 @@
+            clearOldestOverLimitStandard(context, notifsToMakeRoomFor);
+         else
+            clearOldestOverLimitFallback(context, notifsToMakeRoomFor);
+      } catch(Throwable t) {
+         // try-catch for Android 6.0.X bug work around, getActiveNotifications bug
+         clearOldestOverLimitFallback(context, notifsToMakeRoomFor);
+      }
+   }
+
+   // Cancel the oldest notifications based on what the Android system reports is in the shade.
+   // This could be any notification, not just a OneSignal notification
+   @RequiresApi(api = Build.VERSION_CODES.M)
+   static void clearOldestOverLimitStandard(Context context, int notifsToMakeRoomFor) throws Throwable {
+      NotificationManager notifManager = (NotificationManager)context.getSystemService(Context.NOTIFICATION_SERVICE);
+      StatusBarNotification[] activeNotifs = notifManager.getActiveNotifications();
+
+      int notifsToClear = (activeNotifs.length - getMaxNumberOfNotificationsInt()) + notifsToMakeRoomFor;
+      // We have enough room in the notification shade, no need to clear any notifications
+      if (notifsToClear < 1)
+         return;
+
+      // Create SortedMap so we can sort notifications based on display time
+      SortedMap<Long, Integer> activeNotifIds = new TreeMap<>();
+      for (StatusBarNotification activeNotif : activeNotifs) {
+         if (isGroupSummary(activeNotif))
+            continue;
+         activeNotifIds.put(activeNotif.getNotification().when, activeNotif.getId());
+      }
+
+      // Clear the oldest based on the count in notifsToClear
+      for(Map.Entry<Long, Integer> mapData : activeNotifIds.entrySet()) {
+         OneSignal.cancelNotification(mapData.getValue());
+         if (--notifsToClear <= 0)
+            break;
+      }
+   }
+
+   // This cancels any notifications based on the oldest in the local SQL database
+   static void clearOldestOverLimitFallback(Context context, int notifsToMakeRoomFor) {
+      OneSignalDbHelper dbHelper = OneSignalDbHelper.getInstance(context);
+
+      Cursor cursor = null;
+      try {
+         SQLiteDatabase readableDb = dbHelper.getReadableDbWithRetries();
+         cursor = readableDb.query(
+            NotificationTable.TABLE_NAME,
+            new String[] { NotificationTable.COLUMN_NAME_ANDROID_NOTIFICATION_ID },
+            NotificationTable.recentUninteractedWithNotificationsWhere().toString(),
+            null,
+            null,
+            null,
+            OneSignalDbContract.NotificationTable._ID, // sort order, old to new
+            getMaxNumberOfNotificationsString() + notifsToMakeRoomFor // limit
+         );
+
+         int notifsToClear = (cursor.getCount() - getMaxNumberOfNotificationsInt()) + notifsToMakeRoomFor;
+         // We have enough room in the notification shade, no need to clear any notifications
+         if (notifsToClear < 1)
+            return;
+
+         while (cursor.moveToNext()) {
+            notifsToClear--;
+            if (notifsToClear < 0)
```

Comment:
Small syntax comment, on line 78 you are both decrementing and checking against zero on the same line, but in this loop you are doing these on separate lines. Maybe we could make it consistent in both places